torch/utils/data/datapipes/iter/selecting.py

```diff
@@ -12,28 +12,53 @@ class FilterIterDataPipe(MapIterDataPipe):
 
     Iterable DataPipe to filter elements from datapipe according to filter_fn.
     args:
-        datapipe: Iterable DataPipe being filterd
+        datapipe: Iterable DataPipe being filtered
         filter_fn: Customized function mapping an element to a boolean.
         fn_args: Positional arguments for `filter_fn`
         fn_kwargs: Keyword arguments for `filter_fn`
+        drop_empty_batches: By default, drops batch if it is empty after filtering instead of keeping an empty list
+        nesting_level: Determines which level the fn gets applied to, by default it applies to the top level (= 0)
     """
     def __init__(self,
                  datapipe: IterDataPipe[T_co],
                  filter_fn: Callable[..., bool],
                  fn_args: Optional[Tuple] = None,
                  fn_kwargs: Optional[Dict] = None,
+                 drop_empty_batches: bool = True,
+                 nesting_level: int = 0,
                  ) -> None:
-        super().__init__(datapipe, fn=filter_fn, fn_args=fn_args, fn_kwargs=fn_kwargs)
+        self.drop_empty_batches = drop_empty_batches
+        super().__init__(datapipe, fn=filter_fn, fn_args=fn_args, fn_kwargs=fn_kwargs, nesting_level=nesting_level)
+
+    def _merge(self, data, mask):
+        result = []
+        for i, b in zip(data, mask):
+            if isinstance(b, list):
+                t = self._merge(i, b)
+                if len(t) > 0 or not self.drop_empty_batches:
+                    result.append(t)
+            else:
+                if b:
+                    result.append(i)
+        return result
 
     def __iter__(self) -> Iterator[T_co]:
         res: bool
         for data in self.datapipe:
-            res = self.fn(data, *self.args, **self.kwargs)
-            if not isinstance(res, bool):
-                raise ValueError("Boolean output is required for "
-                                 "`filter_fn` of FilterIterDataPipe")
-            if res:
-                yield data
+            if self.nesting_level > 0 and not isinstance(data, list):
+                raise IndexError(f"nesting_level {self.nesting_level} out of range (exceeds data pipe depth)")
+            elif self.nesting_level == 0 or not isinstance(data, list):
```

Comment:
Ahhh, we change the case of `nesting_level > 0` and data is not list. But when `nesting_level == 0`, why data is required to be list? I would assume to run `filter_fn` directly run over any data in this case.

So, the basic logic here should be:
1. self.nesting_level == 0 -> apply `filter_fn` and return
2. Other cases, call `_apply` and `_merge`. (Let `_apply` to handle the `IndexError`)

Please let me know if it makes sense. We can discuss about it.

Reply:
Yes, I think this is a lot cleaner.